gateway-messages/src/ereport.rs

```diff
@@ -0,0 +1,383 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2025 Oxide Computer Company
+
+//! Request and response types for the ereport ingestion protocol.
+//!
+//! This module defines both the request message sent by MGS
+//! ([`EreportRequest`]), and the header for the response to that message by the
+//! SP ([`EreportResponseHeader`]).
+//!
+//! For more information on the high-level design of the ereport ingestion
+//! protocol, see [RFD 520] Control Plane Fault Ingestion and Data Model. For
+//! details on the encoding of SP ereport messages, refer to [RFD 544] Embedded
+//! E-report Formats and [RFD 545] Firmware E-report Aggregation and Evacuation.
+//!
+//! [RFD 520] https://rfd.shared.oxide.computer/rfd/520
+//! [RFD 544] https://rfd.shared.oxide.computer/rfd/544
+//! [RFD 545] https://rfd.shared.oxide.computer/rfd/545
+
+use core::fmt;
+use hubpack::SerializedSize;
+use serde::Deserialize;
+use serde::Serialize;
+
+/// An error numeric identifier (ENA).
+///
+/// ENAs are 64-bit unsigned integers that uniquely identify an ereport within a
+/// single reporter restart.
+///
+/// See [RFD 520 §1.1.1] for details.
+///
+/// [RFD 520 §1.1.1]: https://rfd.shared.oxide.computer/rfd/0520#enas
+#[derive(
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    Ord,
+    PartialOrd,
+    Serialize,
+    Deserialize,
+    SerializedSize,
+)]
+#[repr(transparent)]
+pub struct Ena(pub u64);
+
+impl fmt::Debug for Ena {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Ena({:#x})", self.0)
+    }
+}
+
+/// A unique identifier generated by the SP's snitch task when it starts up,
+/// used to detect restarts.
+///
+/// See [RFD 520 §4.2.2] for details.
+///
+/// [RFD 520 §4.2.2]: https://rfd.shared.oxide.computer/rfd/0520#reporter-crash-recovery
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, SerializedSize,
+)]
+#[repr(transparent)]
+pub struct RestartId(pub u128);
```

Comment:
I did consider using the `uuid` crate for this, as it supports `no_std`, but it didn't really seem worth adding another dependency that would have to be compiled in to the Hubris binaries basically just to get UUID-like formatting in `Debug` impls that are only used by MGS...